pages/api/arweave/upload.ts

```diff
@@ -0,0 +1,155 @@
+import { ArweaveFile } from '@/modules/arweave/types';
+import {
+  ArweaveUploadParams,
+  ArweaveUploadPayload,
+  FILE_FORM_NAME,
+  formatMessage,
+  PAYLOAD_FORM_NAME,
+} from '@/modules/arweave/upload';
+import {
+  ajvParse,
+  JsonString,
+  parseNotarized,
+  unpackNotarized,
+  verifyNaclSelfContained,
+} from '@/modules/notary';
+import { resultThenAsync } from '@/modules/result';
+import { ApiError, FormData } from '@/modules/utils';
+import { WALLETS } from '@/modules/wallet/server';
+import { PublicKey } from '@solana/web3.js';
+import Ajv, { JTDSchemaType } from 'ajv/dist/jtd';
+import { sha256 } from 'crypto-hash';
+import { File, IncomingForm } from 'formidable';
+import fs from 'fs/promises';
+import { NextApiRequest, NextApiResponse } from 'next';
+
+/** JSON schemas for parsing request parameters. */
+const SCHEMAS = (() => {
+  const ajv = new Ajv();
+
+  const payload: JTDSchemaType<ArweaveUploadPayload> = {
+    properties: {
+      pubkey: { type: 'string' },
+      fileHash: { type: 'string' },
+    },
+  };
+
+  return { parsePayload: ajvParse(ajv.compileParser(payload)) };
+})();
+
+/** Verify a notarized post request, returning the file to upload. */
+const verifyPostParams = async (params: FormData) => {
+  const { parsePayload } = SCHEMAS;
+
+  const notarizedFields: string[] | string | undefined = params.fields[PAYLOAD_FORM_NAME];
+  const files: File[] | File | undefined = params.files[FILE_FORM_NAME];
+
+  if (notarizedFields === undefined || files === undefined)
+    throw new ApiError(400, 'Invalid request parameters');
+
+  const notarized: string = notarizedFields instanceof Array ? notarizedFields[0] : notarizedFields;
+
+  const payloadRes = await resultThenAsync(
+    parseNotarized<ArweaveUploadPayload>(notarized as JsonString<ArweaveUploadParams>),
+    (payload) =>
+      unpackNotarized(
+        payload,
+        verifyNaclSelfContained((s) => new PublicKey(s.pubkey).toBuffer()),
+        { parse: parsePayload, format: formatMessage }
+      )
+  );
```

Comment:
1. Parse the unverified payload to get the pubkey that should be the signer.
2. Pass pubkey to verfiy. 
3. Reject or accept based on the signature being from the pub key in the payload.